src/models/poll.ts

```diff
@@ -0,0 +1,178 @@
+/*
+Copyright 2022 The Matrix.org Foundation C.I.C.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+import { M_POLL_END, M_POLL_RESPONSE, PollStartEvent } from "matrix-events-sdk";
+
+import { MatrixClient } from "../client";
+import { MatrixEvent } from "./event";
+import { Relations } from "./relations";
+import { TypedEventEmitter } from "./typed-event-emitter";
+
+export enum PollEvent {
+    New = "Poll.new",
+    End = "Poll.end",
+    Update = "Poll.update",
+    Responses = "Poll.Responses",
+    Destroy = "Poll.Destroy",
+}
+
+export type PollEventHandlerMap = {
+    [PollEvent.Update]: (event: MatrixEvent, poll: Poll) => void;
+    [PollEvent.Destroy]: (pollIdentifier: string) => void;
+    [PollEvent.End]: () => void;
+    [PollEvent.Responses]: (responses: Relations) => void;
+};
+
+const filterResponseRelations = (
+    relationEvents: MatrixEvent[],
+    pollEndTimestamp: number,
+): {
+    responseEvents: MatrixEvent[];
+} => {
+    const responseEvents = relationEvents.filter((event) => {
+        if (event.isDecryptionFailure()) {
+            // @TODO(kerrya) PSG-1023 track and return these
+            return;
+        }
+        return (
+            M_POLL_RESPONSE.matches(event.getType()) &&
+            // From MSC3381:
+            // "Votes sent on or before the end event's timestamp are valid votes"
+            event.getTs() <= pollEndTimestamp
+        );
+    });
+
+    return { responseEvents };
+};
+
+// https://github.com/matrix-org/matrix-spec-proposals/pull/3672
+export class Poll extends TypedEventEmitter<Exclude<PollEvent, PollEvent.New>, PollEventHandlerMap> {
+    public readonly roomId: string;
+    private pollEvent: PollStartEvent | undefined;
+    private fetchingResponsesPromise: null | Promise<void> = null;
+    private responses: null | Relations = null;
+    private endEvent: MatrixEvent | undefined;
+
+    public constructor(private rootEvent: MatrixEvent, private matrixClient: MatrixClient) {
+        super();
+        this.roomId = this.rootEvent.getRoomId()!;
+        this.setPollStartEvent(this.rootEvent);
+    }
+
+    public get pollId(): string {
+        return this.rootEvent.getId()!;
+    }
+
+    public get isEnded(): boolean {
+        return !!this.endEvent;
+    }
+
+    public setPollStartEvent(event: MatrixEvent): void {
+        this.pollEvent = event.unstableExtensibleEvent as PollStartEvent;
+    }
+
+    public getPollStartEvent(): PollStartEvent {
+        return this.pollEvent!;
```

Comment:
Had the same „issue“ with all these `!` while working on broadcasts. What I did is not setting the value in a method. Instead I changed it to return the value and assigned it inside the ctor.

Here this would mean something like `setPollStartEvent` → `determinePollStartEvent` and inside the ctor `this.pollEvent = this.determinePollStartEvent(…)`.

Reply:
Good point, I've replaced this with a `public readonly pollEvent` that's assigned in the constructor.